cuda/btx_cudainterval_callbacks.cpp

```diff
@@ -0,0 +1,589 @@
+                                       const char *hostname, int64_t vpid,
+                                       uint64_t vtid, CUresult cuResult,
+                                       CUcontext ctx) {
+  auto state = static_cast<data_t *>(usr_data);
+  state->context_manager.ctx_pop_current_exit({hostname, vpid, vtid}, cuResult, ctx);
+}
+
+void stream_create_exit_callback(void *btx_handle, void *usr_data,
+                                 int64_t ts, const char *event_class_name,
+                                 const char *hostname, int64_t vpid,
+                                 uint64_t vtid, CUresult cuResult,
+                                 CUstream stream) {
+  auto state = static_cast<data_t *>(usr_data);
+  hpt_t hpt = {hostname, vpid, vtid};
+  state->context_manager.stream_create_exit({hostname, vpid, vtid}, cuResult, stream);
+}
+// ===============================
+// END Context management
+// ===============================
+
+// ===============================
+// Kernel name and device tracking
+// ===============================
+void module_get_function_entry_callback(void *btx_handle, void *usr_data,
+                                        int64_t ts, const char *event_class_name,
+                                        const char *hostname, int64_t vpid,
+                                        uint64_t vtid, char* name) {
+  std::string name_str(name);
+  auto state = static_cast<data_t *>(usr_data);
+  hpt_t hpt = {hostname, vpid, vtid};
+  state->entry_kernel_name[hpt] = name_str;
+}
+
+void module_get_function_exit_callback(void *btx_handle, void *usr_data,
+                                       int64_t ts, const char *event_class_name,
+                                       const char *hostname, int64_t vpid,
+                                       uint64_t vtid, CUresult cuResult,
+                                       CUfunction cuFunction) {
+  if (cuResultIsError(cuResult)) {
+    return;
+  }
+  auto state = static_cast<data_t *>(usr_data);
+  hpt_t hpt = {hostname, vpid, vtid};
+  try {
+    auto kernel_name = state->entry_kernel_name.at(hpt);
+    hp_kernel_t hp_kernel_key = {hostname, vpid, cuFunction};
+    state->hp_kernel_to_name[hp_kernel_key] = kernel_name;
+  } catch(const std::out_of_range& oor) {
+    THAPI_FATAL_HPT(hpt,
+                    "entry kernel name not found in module get function exit callback");
+  }
+}
+
+void operation_entry_helper(void *btx_handle, void *usr_data,
+                            int64_t ts, const char *event_class_name,
+                            const char *hostname, int64_t vpid,
+                            uint64_t vtid, CUstream cuStream, CUfunction f) {
+  auto state = static_cast<data_t *>(usr_data);
+  hpt_t hpt = {hostname, vpid, vtid};
+  CUdevice dev;
+  bool found = state->context_manager.get_stream_device(hpt, cuStream, &dev);
+  if (!found) {
+    THAPI_FATAL_HPT(hpt, "no device for thread in device op");
+  }
```

Comment:
```suggestion
  bool found = state->context_manager.get_stream_device(hpt, cuStream, &dev);
  if (!found) {
    THAPI_FATAL_HPT(hpt, "no device for thread in device op");
  }
```
```suggestion
 if (state->context_manager.get_stream_device(hpt, cuStream, &dev)
    THAPI_FATAL_HPT(hpt, "no device for thread in device op");
```

Reply:
Makes line too long to put inside the if, and I like the clarity of found name.

Reply:
Also this is the one place I left THAPI_FATAL. Not sure if worth keeping the macros or keeping this check. May need a perf check to see if having context manager check this slows it down at all. Context manager is at least called a lot less then e.g. entry state, i.e. on per task basis, not on per callback bases. And code that does lots of small kernel launches is generally discouraged.

Reply:
If the logic of our code is correct, this should never happen.

If this happend, I guess we can always put back some tests, (or wait for the `.at()` improvement / debug)